gcp/app-engine/main.py

```diff
@@ -141,6 +141,47 @@ def get_pga_player_scorecards(user_email):
         records = db_conn.execute(sqlalchemy.text(query)).fetchall()
         return json.dumps([row[0] for row in records], default=json_serial), 200
 
+@app.route(PUBLIC_API_PREFIX + '/users', methods=['POST'])
+@validate_token
+def post_user(user_email):
+    print(f"Got post_user request from user: {user_email}")
+    data = request.get_json(force=True)
+    if data is None:
+        return 'Post body empty', 400
+    if 'email' not in data:
+        return 'Post body must include email', 400
+
+    email = data.get('email').replace("'", "''")
+    id = data.get('id')
+    name = data.get('name')
+    now = datetime.utcnow()
+    if id:
+        id = id.replace("'", "''")
+    if name:
+        name = name.replace("'", "''")
+
+    on_conflict_updates = [
+        f"email = EXCLUDED.email",
+        f"last_login = '{str(now)}'",
+        # Use the actual name instead of EXCLUDED.name since EXCLUDED.name could be the email
+        f"last_updated = (case when users.email != EXCLUDED.email or users.name != '{name}' then EXCLUDED.last_updated else users.last_updated end)"
+    ]
+    if name:
+        # Use the actual name instead of EXCLUDED.name since EXCLUDED.name could be the email
+        on_conflict_updates.append(f"name = '{name}'")
+
+    pool = get_db_connection()
+    with pool.connect() as db_conn:
+        db_conn.execute(sqlalchemy.text(f'''
+            insert into users (id, name, email, created, last_updated, last_login)
```

Comment:
insert the user info if it doesn't exist, otherwise update last_login and optionally the name+email+last_updated if modified

Reply:
Note this has already been pushed to the app engine, so you should be able to run this code without using a local app engine

Reply:
Also we should create a ticket to use the sqlalchemy query builder for this rather than raw sql strings.